### PR TITLE
Support for non-immediate proof maintenance

### PR DESCRIPTION
**Pull request by PetarMax**
When doing large real-world proofs with a high degree of parallelism, the bandwidth is overtaken by the writing of the proof to disk on every proof iteration, slowing the process considerably.

This PR introduces the ability to write to disk and perform the callback on every `maintenance_iterations` of the parallel proof, set by default to 32.

EDIT: Actually, perhaps it's a better idea to write the proof every `max_workers` iterations.

### CODE REVIEW

**Review by nwatson22 — APPROVED** [review]
Looks good. Can you make it toggleable with a flag so we can choose to do the disk writing/status bar updating every step?

**Comment by PetarMax** [comment]
I think I prefer to have a separate parameter so that we can decouple it from the number of workers, as it is now. Is there anything else to do in `pyk`?

**Comment by nwatson22** [comment]
> I think I prefer to have a separate parameter so that we can decouple it from the number of workers, as it is now. Is there anything else to do in `pyk`?

You could possibly add the same parameter/logic to `Prover.advance_proof()` and feed the option through to be able to be used with `pyk prove`, but that could also be done later in a separate PR since this is needed for kontrol now.

File: pyk/src/tests/unit/test-data/pyk_toml_test.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [coverage]
 output = "default-file"
-definition = "/tmp"
+definition = "/var/folders/ks/1p8zyp5j7xz_v1krhl4l17tm0000gn/T"
```
Comment (nwatson22) [inline pyk/src/tests/unit/test-data/pyk_toml_test.toml]: What is this change for?

**Review by nwatson22 — APPROVED** [review]

File: pyk/src/tests/unit/test-data/pyk_toml_test.toml (hunk 1 above)
Comment (PetarMax) [reply pyk/src/tests/unit/test-data/pyk_toml_test.toml]: I have not changed that file manually...

**Comment by PetarMax** [comment]
> > I think I prefer to have a separate parameter so that we can decouple it from the number of workers, as it is now. Is there anything else to do in `pyk`?
> 
> You could possibly add the same parameter/logic to `Prover.advance_proof()` and feed the option through to be able to be used with `pyk prove`, but that could also be done later in a separate PR since this is needed for kontrol now.

Yes, happy to.

**Comment by PetarMax** [comment]
Just to say - I am not sure how to test this, because the final outcome is the same, it's just that the proof should be faster. Happy to go along with any ideas in a follow-up PR.

**Pull request merged**